### ISSUE TITLE
wx.GraphicsContext.SetCompositionMode not working correctly?

### ISSUE THREAD

**Issue by kdschlosser** [wxWidgets/Phoenix#1810]
**Operating system**: Windows 7 X64
**wxPython version & source**: 4.0.7.post1 msw (phoenix) wxWidgets 3.0.5 (pypi)
**Python version & source**:  CPython 3.7.5

**Description of the problem**:
I am not sure if there is a problem or not, I know that the QT framework does support the modes that would be considered equivalent when wx reports that they are unsupported. The documentation does not go into detail about setting the composition mode or what the various modes exactly do except for a math equation that has no breakdown on what the variables are in the equation. 

I guess I do not understand why there is no support for modes if the mode is just a math equation that would get used against the pixels being drawn and the pixels that already exist on the screen. There is no explanation as to why a mode would be unsupported, is it the OS? is it the video card or is it the DC type being used?

Here is the list of supported modes on my system
COMPOSITION_SOURCE
COMPOSITION_DEST
COMPOSITION_OVER

Here is the list of unsupported modes
COMPOSITION_CLEAR
COMPOSITION_IN
COMPOSITION_OUT
COMPOSITION_ATOP
COMPOSITION_DEST_OVER
COMPOSITION_DEST_IN
COMPOSITION_DEST_OUT
COMPOSITION_DEST_ATOP
COMPOSITION_XOR
COMPOSITION_ADD

One would think that COMPOSITION_ADD would be supported no matter what because the rendering of a color that has an alpha channel on top of another color would add the 2 together correct?

**Comment by RobinD42** [comment]
It depends on what modes are supported by the platform-specific APIs used to implement the GC classes. One way to get all the composition modes on Windows is to use the Cairo backend for the GC classes instead of the default GDI+. There is some code in the demo that shows how to do that.

**Comment by kdschlosser** [comment]
If I use Cairo is it possible to "erase" what has been drawn??.. I am working on a project where there are changes that if I had a way to erase them it would make updating a lot faster. Right now if I have to remove something I have to render the whole thing over again minus the pieces that need to be removed.

I have a data stream that runs at 33.3kbits a second, the data packet is 104 bits. speed if of greatest importance and the most I can am able to get out of it the better it is going to be. I am now buffering all of the updates to keep the performance up. The issue is it comes with an extremely high memory use which I am not exactly why it is using so much memory, 1000mb of ram to load 9.82mb of png files as wxBitmaps.

**Comment by RobinD42** [comment]
No, the graphics context doesn't operate at that high of a level for any backend. It does not track objects that have been drawn or offer any way to manipulate them. Even if it did, it would still need to redraw everything, or at least everything in the area being modified, in order to make changes.

There are a few approaches to help improve performance of drawing operations. 

* When something changes in the data that would cause a change in the display then use `RefreshRect` with the rectangle corresponding to the area in the window that needs redrawn to show that change. Then in the `EVT_PAINT` event use the update region (which will contain that rectangle plus any others that have been added since the last paint event) and only redraw the items that intersect with the region.

* If your display can be partitioned into static things, seldom changing things, and frequently changing things, and they can be displayed as layers, then you can draw the static things just one time to a bitmap, and the seldom changing things to another bitmap when they change, and then your paint event can be reduced to drawing the static and seldom bitmaps followed by the frequently changing drawing operations.

* Something like the `wx.adv.PseudoDC` for `GraphicsContexts` would be a nice addition. So would a higher level canvas class like `FloatCanvas` for GCs too, although that would probably end up not being any help in a high performance high data throughput scenario like you're dealing with.

**Comment by kdschlosser** [comment]
what I have done is I am using frames for each "layer" I am using UpdateLayeredWindow and passing a BLENDFUNCTION structure to it to render a completely transparent frame. I am doing what you have suggested already and rendering to a bitmap. I then pass the handle for that bitmap to UpdateLayeredWindow to render the bitmap. Each piece that makes up the whole image I am rendering has it's own frame and can be updated if needs be separately without effecting any of the other layers. This does work and it works well. Because of the type of information I am changing adding new pixel data is fast, this is ebcause the pixel data is a small amount. If I had a way to erase a small amount it would be just as fast. Unfortunately the only way I have found to do that is to redraw the entire bitmap over again without the data that I want deleted. I am not sure why I am having such a performance hit doing this it is not a complex thing to do. There is no function in GraphicsContext like there is in GCDC to pass a list of lines and panes to be drawn. I am sure that this would be faster then having to iterate over the list calling StrokeLine for each line I want to draw.

So I then tried buffering the bitmaps I have already rendered so I can skip the rendering of them. This has added a HUGE speed boost but at a large expense. massive amounts of memory used for an extremely small amount of data being loaded. I am not sure why this is. But over 1000 mb of ram used when loading < 10mb of bitmaps doesn't seem correct.... I don't know if there is some kind of a memory leak somewhere in wxBitmap. I also posted an issue about a strange internal application crash that is happening as well.

If the massive consumption of memory was fixed then I have that high level GraphicsContext layer design that you outlined above, just doing it in a really different manner.

**Comment by Metallicow** [comment]
dc.DrawBitmap( is the usually e slowest one of the dc. Draw methods most times.
It would be EXTREMELY nice to have multiple input versions of all the dc.Draw methods.
For example: 
DrawEllipseList | Draw a list of ellipses as quickly as possible.
already exists.
We really need a
DrawBitmapList method for performance done in the C side of things. 
otherwise you will continue to see me/others use list comprehensions or complex generators to try and draw as fast as we can.

@RobinD42 
These methods have a multiple 'List' way of doing is relatively fast.
```
DrawEllipse
DrawEllipseList

DrawLine
DrawLineList
DrawLines

DrawPoint
DrawPointList

DrawPolygon
DrawPolygonList

DrawRectangle
DrawRectangleList

DrawText
DrawTextList
```

These methods don't have an optimized 'List' method on the C side.
```
DrawArc
DrawBitmap
DrawCheckMark
DrawCircle
DrawEllipticArc
DrawIcon
DrawLabel
DrawRotatedText
DrawRoundedRectangle
DrawSpline
```

**Comment by kdschlosser** [comment]
@Metallicow 
You are referring to DC and any DC subclasses. This issue and conversation is not about DC or any of the subclasses. It is about GraphicsContext which does not have the same methods. I am not using DrawBitmap at all not even in the Graphics context. I am using the UpdateLayeredWindow function of the Windows api to render the bitmap, this is by far the fastest way I have come up with to render. Because what I am rendering is something that I am able to buffer that is what I am doing. 

The buffering leads to a problem of memory consumption, HUGE amounts of ram used for a small amount of data. I don't know why it is doing this. as I stated above loading <10mb worth of bitmaps that get loaded using wx.Bitmap consumes over a gig of ram. There has got to be some kind of a memory issue with wxBitmap or wx.Bitmap somewhere.

**Comment by Metallicow** [comment]
Ahhh Ok. I figured maybe you might have played around with the get/set graphics context while trying to speed things up with a GCDC.
I got graphic context methods for all that stuff but I implemented them all in python. so yea, it is a bit slower than the c stuff and uses a lot of comprehensions/generators and some might take slightly bit different args than the dc equivalents. plus the gc stuff is sometimes offset by 0.5 so its a bit of a headache to get everything fine tuned for the matrix.

Not sure about why it is taking so much memory... probably bitmap or gc's implementation dealing with them..

Yea a PseudoDC for graphics contexts would definitely be nice in C. I thought of that also. You could write one but it would be a pure python impl so it definitely wouldn't perform the way you are wanting at animation speeds and it would take a good bit of time to reimplement all the gcdc stuff in pure py.
